### PR TITLE
LKE-5884 feat (autoRange): Handle logarithmic function

### PR DESCRIPTION
**Pull request by Kraamel**
We want to be able to use the autoRange with either a linear (default) or a logarithmic function

Spec: https://www.notion.so/linkurious/Dynamic-Sizing-a1c3c682647d407fbeb97b3541fc9f45#487ade2c43284e0b93ecc68e5e2e72f3

### CODE REVIEW

File: src/styles/edgeAttributes.ts [hunk 1]
```diff
@@ -82,7 +82,10 @@ export class EdgeAttributes extends ItemAttributes {
         ) {
           const propertyName: string = widthStyle.input[1];
           const propertyValue = Tools.parseNumber(data.properties[propertyName]);
-          return EdgeAttributes.getAutomaticRangeWidth(propertyValue, styleRule);
+          //to update with the correct enum type
+          return widthStyle.scale && widthStyle.scale === 'logarithmic'
```
Comment (aminerhanemi) [inline src/styles/edgeAttributes.ts:86]: This three lines can be replaced with.
```
const isLog = widthStyle.scale && widthStyle.scale === 'logarithmic'
return EdgeAttributes.getAutomaticRangeWidth(propertyValue, styleRule, isLog)
```

What do you think ?

File: src/styles/itemAttributes.ts [hunk 2]
```diff
@@ -141,4 +141,35 @@ export class ItemAttributes {
...
+   */
+  public static getAutomaticRangeStyleLog(
+    value: number,
+    {max, min}: {max: number; min: number},
+    lower: EdgeWidthExtrema | NodeSizeExtrema,
+    higher: EdgeWidthExtrema | NodeSizeExtrema
+  ): string {
+    // apply an offset for all the values (including max) if min is smaller than one and set min to 1
+    if (min < 1) {
+      value += -min + 1;
```
Comment (aminerhanemi) [inline src/styles/itemAttributes.ts:160]: it is better to use absolute value than -min in case min is between 0 and 1.
```suggestion
      value += Math.abs(min)+ 1;
```

File: src/styles/itemAttributes.ts [hunk 3]
```diff
@@ -141,4 +141,35 @@ export class ItemAttributes {
...
+      max += -min + 1;
+      min = 1;
```
Comment (aminerhanemi) [inline src/styles/itemAttributes.ts:162]: same

File: src/styles/nodeAttributes.ts [hunk 4]
```diff
@@ -32,7 +32,7 @@ export class NodeAttributes extends ItemAttributes {
   /**
    * Run the callback if an item match with a style in the array of rules
    */
-  private matchStyle(
+  private static matchStyle(
```
Comment (aminerhanemi) [inline src/styles/nodeAttributes.ts:35]: unit test if it is possible

File: src/styles/nodeAttributes.ts [hunk 5]
```diff
@@ -181,7 +181,11 @@ export class NodeAttributes extends ItemAttributes {
           ) {
             const propertyName: string = sizeStyle.input[1];
             const propertyValue = Tools.parseNumber(itemData.properties[propertyName]);
-            result = NodeAttributes.getAutomaticRangeSize(propertyValue, styleRule);
+            //to update with the correct enum type
+            result =
```
Comment (aminerhanemi) [inline src/styles/nodeAttributes.ts:185]: Same comment as edges

File: src/styles/itemAttributes.ts [hunk 6]
```diff
@@ -155,16 +155,16 @@ export class ItemAttributes {
...
-    if (min < 1) {
-      value += -min + 1;
-      max += -min + 1;
-      min = 1;
-    }
     // apply default style when min equal max
     if (max === min) {
       return '100%';
     }
+    // apply an offset for all the values (including max) if min is smaller than one and set min to 1
```
Comment (aminerhanemi) [inline src/styles/itemAttributes.ts:162]: // apply an offset for all the values (including min and max)

File: tests/styles/edgeAttributes.spec.ts [hunk 7]
```diff
@@ -414,6 +414,29 @@ describe('EdgeAttributes', () => {
           })
         )
       ).to.equal('50%');
+
+      expect(
```
Comment (aminerhanemi) [inline tests/styles/edgeAttributes.spec.ts:418]: can you add test with negative value

File: tests/styles/nodeAttributes.spec.ts [hunk 8]
```diff
@@ -4928,6 +4928,29 @@ describe('NodeAttributes', function () {
           })
         )
       ).to.equal('93%');
+
+      expect(
+        NodeAttributes.getAutomaticRangeSize(
```
Comment (aminerhanemi) [inline tests/styles/nodeAttributes.spec.ts:4933]: same

**Review by aminerhanemi — CHANGES_REQUESTED** [review]

File: tests/styles/edgeAttributes.spec.ts (hunk 7 above)
Comment (Kraamel) [reply tests/styles/edgeAttributes.spec.ts:418]: done

File: tests/styles/nodeAttributes.spec.ts (hunk 8 above)
Comment (Kraamel) [reply tests/styles/nodeAttributes.spec.ts:4933]: done

**Review by aminerhanemi — APPROVED** [review]
